# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/workers/WorkerPrivate.cpp
# Commit: 893865b3c353
# Full Hash: 893865b3c3532d57c92570ae0e7e85469bccc444
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2024-07-03 16:15:19
# Description:
#   Bug 1905337 - Keep dispatched RefPtr<WorkerThreadRunnable> of WorkerPrivate::mPreStartRunnables. r=dom-worker-reviewers,jstutte,smaug
#   
#   https://phabricator.services.mozilla.com/D213947 introduced a post-handling for WorkerPrivate::mPreStartRunnables once Worker initialization fails.
#   However, it needs corresponding WorkerThreadRunnable be kept in WorkerPrivate::mPreStartRunnables. However, [[ https://searchfox.org/mozilla-central/rev/3d173a6ad865eb778eb7a85de900e92774559ed6/dom/workers/WorkerPrivate.cpp#5928 | we did not keep it while dispatching]], and then meet UAF when executing the post-handling.
#   
# ==============================================================================

diff -r f778b37ce90f -r 893865b3c353 dom/workers/WorkerPrivate.cpp
--- a/dom/workers/WorkerPrivate.cpp	Wed Jul 03 09:32:28 2024 +0000
+++ b/dom/workers/WorkerPrivate.cpp	Wed Jul 03 09:59:16 2024 +0000
@@ -5925,7 +5925,7 @@
   if (!mPreStartRunnables.IsEmpty()) {
     for (uint32_t index = 0; index < mPreStartRunnables.Length(); index++) {
       MOZ_ALWAYS_SUCCEEDS(mThread->DispatchAnyThread(
-          WorkerThreadFriendKey{}, mPreStartRunnables[index].forget()));
+          WorkerThreadFriendKey{}, mPreStartRunnables[index]));
     }
     // Don't clear mPreStartRunnables here, it will be cleared in the beginning
     // of WorkerPrivate::DoRunLoop() or when in WorkerPrivate::RunLoopNeverRan()